include/verilated_vpi.cpp

```diff
@@ -2812,13 +2824,1086 @@ vpiHandle vpi_put_value(vpiHandle object, p_vpi_value valuep, p_vpi_time /*time_
+        PLI_INT32 *integers;
+
+        if (arrayvalue_p->flags & vpiUserAllocFlag) {
+            integers = arrayvalue_p->value.integers;
+        } else {
+            if (VL_UNCOVERABLE(num >= VL_VALUE_STRING_MAX_WORDS)) {
+                VL_FATAL_MT(__FILE__, __LINE__, "",
+                            "vpi_get_value_array with more than VL_VALUE_STRING_MAX_WORDS; "
+                            "increase and recompile");
+            }
+
+            integers = (PLI_INT32*)out_ptr;
+            arrayvalue_p->value.integers = integers;
+        }
+
+        if (varp->vltype() == VLVT_UINT8) {
+            CData *ptr = reinterpret_cast<CData*>(vop->varDatap());
+
+            if (vop->rangep()->left() < vop->rangep()->right()) {
+                for (int i = 0; i < num; i++) {
+                    integers[i] = ptr[index];
+                    index = ++index % size;
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    integers[i] = ptr[index];
+                    index = (--index + size) % size;
+                }
+            }
+        } else if (varp->vltype() == VLVT_UINT16) {
+            SData *ptr = reinterpret_cast<SData*>(vop->varDatap());
+
+            if (vop->rangep()->left() < vop->rangep()->right()) {
+                for (int i = 0; i < num; i++) {
+                    integers[i] = ptr[index];
+                    index = ++index % size;
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    integers[i] = ptr[index];
+                    index = (--index + size) % size;
+                }
+            }
+        } else if (varp->vltype() == VLVT_UINT32) {
+            IData *ptr = reinterpret_cast<IData*>(vop->varDatap());
+
+            if (vop->rangep()->left() < vop->rangep()->right()) {
+                for (int i = 0; i < num; i++) {
+                    integers[i] = ptr[index];
+                    index = ++index % size;
+                }
+            } else {
+                for (int i = 0; i < num; i++) {
+                    integers[i] = ptr[index];
+                    index = (--index + size) % size;
+                }
+            }
+        }
+
+        return;
+    } else if (arrayvalue_p->format == vpiShortIntVal) {
+        PLI_INT16 *shortints;
+
+        if (arrayvalue_p->flags & vpiUserAllocFlag) {
+            shortints = arrayvalue_p->value.shortints;
+        } else {
+            const int words = VL_WORDS_I(num * 16);
+            if (VL_UNCOVERABLE(words >= VL_VALUE_STRING_MAX_WORDS)) {
+                VL_FATAL_MT(__FILE__, __LINE__, "",
+                            "vpi_get_value_array with more than VL_VALUE_STRING_MAX_WORDS; "
+                            "increase and recompile");
+            }
+
+            shortints = (PLI_INT16*)out_ptr;
+            arrayvalue_p->value.shortints = shortints;
+        }
+
+        if (varp->vltype() == VLVT_UINT8) {
```

Comment:
There's a lot of cases in this file, please make sure the tests cover them all (using e.g. code coverage to check, or whatever else). 

Also I think most of the duplicated code would be much more obvious if it called some appropriate C++ template code passing the data type as a template parameter.